modules/navier_stokes/src/executioners/SIMPLE.C

```diff
@@ -0,0 +1,1106 @@
+  // diagonal by dividing it with the relaxation parameter. So the new diagonal will be:
+  // D* = 1/lambda*max(|D|,sum(|Offdiag|))
+  // For more information see
+  //
+  // Juretic, Franjo. Error analysis in finite volume CFD. Diss.
+  // Imperial College London (University of London), 2005.
+  //
+  // The trickery comes with storing everything in the diff-diagonal vector
+  // to avoid the allocation and manipulation of a third vector
+  for (const auto row_i : make_range(matrix.row_start(), matrix.row_stop()))
+  {
+    std::vector<numeric_index_type> indices;
+    std::vector<Real> values;
+    matrix.get_row(row_i, indices, values);
+    Real abs_sum = std::accumulate(
+        values.cbegin(), values.cend(), 0.0, [](Real a, Real b) { return a + std::abs(b); });
+    Real abs_diagonal = std::abs(diff_diagonal(row_i));
+    Real new_diagonal = inverse_relaxation * std::max(abs_sum - abs_diagonal, abs_diagonal);
+    diff_diagonal.set(row_i, new_diagonal);
+  }
+
+  // Time to modify the diagonal of the matrix
+  for (const auto row_i : make_range(matrix.row_start(), matrix.row_stop()))
+    matrix.set(row_i, row_i, diff_diagonal(row_i));
+  matrix.close();
+
+  // Finally, we can create (D*-D) vector which is used for the relaxation of the
+  // right hand side later
+  diff_diagonal.add(-1.0, *original_diagonal);
+  diff_diagonal.close();
+}
+
+void
+SIMPLE::relaxRightHandSide(NumericVector<Number> & rhs,
+                           const NumericVector<Number> & solution,
+                           const NumericVector<Number> & diff_diagonal)
+{
+
+  // We need a working vector here to make sure we don't modify the
+  // (D*-D) vector
+  auto working_vector = diff_diagonal.clone();
+  working_vector->pointwise_mult(solution, *working_vector);
+
+  // The correction to the right hand side is just
+  // (D*-D)*old_solution
+  // For more information see
+  //
+  // Juretic, Franjo. Error analysis in finite volume CFD. Diss.
+  // Imperial College London (University of London), 2005.
+  rhs.add(*working_vector);
+  rhs.close();
+}
+
+Real
+SIMPLE::computeNormalizationFactor(const NumericVector<Number> & solution,
+                                   const SparseMatrix<Number> & mat,
+                                   const NumericVector<Number> & rhs)
+{
+  // This function is based on the description provided here:
+  // @article{greenshields2022notes,
+  // title={Notes on computational fluid dynamics: General principles},
+  // author={Greenshields, Christopher J and Weller, Henry G},
+  // journal={(No Title)},
+  // year={2022}
+  // }
+  // so basically we normalize the residual with the following number:
+  // sum(|Ax-Ax_avg|+|b-Ax_avg|)
+  // where A is the system matrix, b is the system right hand side while x and x_avg are
+  // the solution and average solution vectors
+
+  // We create a vector for Ax_avg and Ax
+  auto A_times_average_solution = solution.zero_clone();
+  auto A_times_solution = solution.zero_clone();
```

Comment:
You could use `auto` with your other clones in this translation unit as well. My recommendation is to just be consistent

Reply:
I have been using auto for the clones, it should be consistent withing SIMPLE and INSFVRCISegregated.

Reply:
Upps never mind, found some which are not!